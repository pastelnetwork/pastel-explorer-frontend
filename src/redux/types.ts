import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';
import { RootState } from '@redux/store';

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export type AppThunkDispatch = ThunkDispatch<RootState, unknown, Action<string>>;
