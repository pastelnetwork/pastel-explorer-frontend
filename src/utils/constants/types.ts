import { LabelKeyObject } from 'react-csv/components/CommonPropTypes';
import { PeriodTypes, TGranularity } from '@utils/helpers/statisticsLib';
import { TStatisticsInfo } from '@utils/types/IStatistics';

export type TThemeColor = {
  name: string;
  backgroundColor: string;
  lineColor?: string;
  splitLineColor: string;
  color: string;
  stack?: string;
  smooth?: boolean;
};

export type TThemeInitOption = {
  theme?: TThemeColor | null;
  data?: number[][];
  dataX?: string[];
  dataY?: number[];
  dataY1?: number[];
  dataY2?: number[];
  chartName: string;
  minY: number;
  maxY: number;
};

export type TLineChartProps = {
  chartName: string;
  period?: PeriodTypes;
  dataX?: string[];
  dataY?: number[];
  dataY1?: number[];
  dataY2?: number[];
  title?: string;
  granularity?: TGranularity;
  info: TStatisticsInfo;
  granularities?: TGranularity[];
  offset: number;
  periods?: PeriodTypes[];
  handleGranularityFilterChange?: (_granularity: TGranularity) => void;
  handleBgColorChange: (_color: string) => void;
  handlePeriodFilterChange?: (_period: PeriodTypes) => void;
};

export type TCsvHeaderType = {
  [key: string]: LabelKeyObject[];
};

export type TScatterChartProps = {
  chartName: string;
  data: number[][];
  dataX: string[];
  title?: string;
  period?: PeriodTypes;
  info: TStatisticsInfo;
  offset: number;
  periods: PeriodTypes[];
  handleBgColorChange: (_color: string) => void;
  handlePeriodFilterChange?: (_period: PeriodTypes) => void;
  handleGranularityFilterChange?: (_granularity: TGranularity) => void;
};
