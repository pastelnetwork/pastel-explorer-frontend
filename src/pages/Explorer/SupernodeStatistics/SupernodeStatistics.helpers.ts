import { HeaderType } from '@components/Table/Table';

import { formatNumber } from '@utils/helpers/formatNumbers/formatNumbers';
import { INetworkSupernodes } from '@utils/types/INetwork';
import { chartColors } from '@components/Charts/DoughnutChart/DoughnutChart.options';
import { translateDropdown } from '@utils/helpers/i18n';

import themeVariant from '@theme/variants';

interface CountryObject {
  [key: string]: number;
}

export interface CountryList {
  totalQuantity: number;
  chartData: {
    headers: Array<string>;
    quantities: Array<number>;
  };
  tableData: {
    name: string;
    quantity: number;
    percentage: string;
  }[];
}

export const headers: Array<HeaderType> = [
  { id: 1, header: 'Country' },
  { id: 2, header: 'Quantity' },
  { id: 3, header: '%' },
];

export const generateChartData = (labels: Array<string>, data: Array<number>) => {
  return {
    labels,
    datasets: [
      {
        data,
        backgroundColor: chartColors,
        borderWidth: 1,
        borderColor: themeVariant.palette.background.paper,
        barThickness: 40,
      },
    ],
  };
};

const generateSortedList = (list: CountryObject) => {
  return Object.entries(list).sort(([, valueA], [, valueB]) => {
    if (valueA < valueB) return 1;
    if (valueA > valueB) return -1;
    return 0;
  });
};

const generateNarrowedCountryList = (list: [string, number][], quantity: number) => {
  return list.map(([country, value]) => {
    return {
      name: country,
      quantity: value,
      percentage: formatNumber((value / quantity) * 100, { decimalsLength: 2 }),
    };
  });
};

const generateOtherCountryList = (list: [string, number][], quantity: number) => {
  const otherCountryQuantity = list.reduce((acc, [, value]) => acc + value, 0);

  return {
    name: translateDropdown('pages.explorer.other'),
    quantity: otherCountryQuantity,
    percentage: formatNumber((otherCountryQuantity / quantity) * 100, { decimalsLength: 2 }),
  };
};

const generateCountryChartData = (tableData: CountryList['tableData']) => {
  return tableData.reduce(
    (acc: { headers: string[]; quantities: number[] }, { name, quantity }) => {
      acc.headers.push(name);
      acc.quantities.push(quantity);
      return acc;
    },
    {
      headers: [],
      quantities: [],
    },
  );
};

const generateSortedNarrowedList = (list: CountryObject, narrowDepth: number) => {
  const countrySortedList = generateSortedList(list);
  const countriesQuantity = countrySortedList.reduce((acc, [, value]) => acc + value, 0);
  const displayCountryList = countrySortedList.slice(0, narrowDepth);
  const otherCountryList = countrySortedList.slice(narrowDepth);

  const countryList = generateNarrowedCountryList(displayCountryList, countriesQuantity);
  const otherList = generateOtherCountryList(otherCountryList, countriesQuantity);

  const tableData = otherList.quantity ? [...countryList, otherList] : countryList;
  const chartData = generateCountryChartData(tableData);

  return {
    totalQuantity: countriesQuantity,
    chartData,
    tableData,
  };
};

export const generateSupernodeCountriesList = (
  nodes: Array<INetworkSupernodes>,
  countryQuantity: number,
): CountryList => {
  const objectList = nodes.reduce((acc: CountryObject, { country }) => {
    return {
      ...acc,
      [country]: acc[country] + 1 || 1,
    };
  }, {});

  return generateSortedNarrowedList(objectList, countryQuantity);
};
