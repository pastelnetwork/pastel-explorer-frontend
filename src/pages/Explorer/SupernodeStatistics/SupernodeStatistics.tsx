import * as React from 'react';

import DoughnutChart from '@components/Charts/DoughnutChart/DoughnutChart';

import { INetworkSupernodes } from '@utils/types/INetwork';

import {
  generateSupernodeCountriesList,
  generateChartData,
  CountryList,
} from './SupernodeStatistics.helpers';
import * as ExplorerMapStyles from '../ExplorerMap/ExplorerMap.styles';

interface SupernodeStatisticsProps {
  supernodes: Array<INetworkSupernodes> | null;
  link?: string;
}

const DISPLAY_COUNTRY_QUANTITY = 5;

const SupernodeStatistics: React.FC<SupernodeStatisticsProps> = ({ supernodes, link = '' }) => {
  const [countryQuantity, setCountryQuantity] = React.useState(0);
  const [countryChartData, setCountryChartData] = React.useState<CountryList['chartData'] | null>(
    null,
  );

  const generateSupernodeCountries = (nodes: Array<INetworkSupernodes>) => {
    const { chartData, totalQuantity } = generateSupernodeCountriesList(
      nodes,
      DISPLAY_COUNTRY_QUANTITY,
    );

    setCountryQuantity(totalQuantity);
    setCountryChartData(chartData);
  };

  React.useEffect(() => {
    if (supernodes) {
      generateSupernodeCountries(supernodes);
    }
  }, [supernodes]);

  const totalSuperNodes = countryChartData?.quantities?.reduce((a, b) => a + b, 0);

  return (
    <ExplorerMapStyles.Container>
      <DoughnutChart
        title="Supernode Statistics"
        innerTitle="Total"
        innerSubtitle={countryQuantity}
        data={
          countryChartData &&
          generateChartData(countryChartData.headers, countryChartData.quantities)
        }
        totalSuperNodes={totalSuperNodes}
        link={link}
      />
    </ExplorerMapStyles.Container>
  );
};

export default SupernodeStatistics;
