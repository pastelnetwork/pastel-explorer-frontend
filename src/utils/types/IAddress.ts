export interface IAddressData {
  amount: number;
  timestamp: number;
  transactionHash: string;
  direction: string;
}

export interface IAddress {
  address: string;
  incomingSum: number;
  outgoingSum: number;
  data: Array<IAddressData>;
}
