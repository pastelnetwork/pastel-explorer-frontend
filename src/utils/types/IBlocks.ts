export interface IBlockTransaction {
  id: string;
  recipientCount: number;
  totalAmount: number;
}

export interface IBlock {
  confirmations: number;
  difficulty: string;
  height: number;
  id: string;
  merkleRoot: string;
  nextBlockHash: string;
  nonce: string;
  previousBlockHash: string;
  size: number;
  solution: string;
  timestamp: number;
  transactionCount: number;
  transactions: Array<IBlockTransaction>;
}

export interface IRawBlock {
  anchor: string;
  bits: string;
  chainwork: string;
  confirmations: number;
  difficulty: string;
  finalsaplingroot: string;
  hash: string;
  height: number;
  merkleroot: string;
  nonce: string;
  previousblockhash: string;
  size: number;
  solution: string;
  time: number;
  transactions: Array<IBlockTransaction>;
  tx: string[];
}
