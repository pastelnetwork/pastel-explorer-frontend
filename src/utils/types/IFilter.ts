export type TFilter = {
  name: string;
  value: string;
};
