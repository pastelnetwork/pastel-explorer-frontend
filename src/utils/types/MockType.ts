import { TFunction } from 'i18next';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type MyMockType = jest.Mock<any, any, any> &
  TFunction<['translation', ...string[]], undefined>;
