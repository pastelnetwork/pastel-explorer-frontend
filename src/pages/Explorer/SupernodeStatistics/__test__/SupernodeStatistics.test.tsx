import { shallow } from 'enzyme';

import 'jest-styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { MyMockType } from '@utils/types/MockType';
import i18next from '../../../../utils/helpers/i18n';
import SupernodeStatistics from '../SupernodeStatistics';

jest.mock('i18next-http-backend');
jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            // noop
          }),
      },
    };
  },
  initReactI18next: {
    type: '3rdParty',
    init: () => {
      // noop
    },
  },
}));
i18next.t = jest.fn().mockImplementation((...arg) => {
  return arg[0];
}) as MyMockType;
const mockStore = configureMockStore();
const store = mockStore({});

describe('pages/SupernodeStatistics', () => {
  const wrapper = shallow(
    <Provider store={store}>
      <SupernodeStatistics />
    </Provider>,
  );

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
