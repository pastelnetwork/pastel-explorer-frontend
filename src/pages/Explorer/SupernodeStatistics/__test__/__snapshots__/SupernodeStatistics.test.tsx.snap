// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/SupernodeStatistics renders correctly 1`] = `ShallowWrapper {}`;
